### PR TITLE
Refactor reserves

### PR DESCRIPTION
**Pull request by cfe316**
This doesn't quite compile yet for technical reasons related to `add_similar_to_expression!`.

I tried refactoring some of the reserve formulations in order to better understand how they should work for fusion, and I think others might appreciate the simplifications.

### CODE REVIEW

**Comment by cfe316** [comment]
This works now. The results for the ISONE_Trizone case (which may be the only standard test case that uses Reserves) are identical when just changing the `hydro_res` formulation. They unfortunately change due to numerical issues with the `storage_all_reserves` formulation, but I've checked that the constraints are the same. For example see commit [aa6b7a6](https://github.com/GenXProject/GenX/pull/562/commits/aa6b7a6ecd58e2d51448b4db0fd5b55f782ac280).

Overall, using `add_similar_to_expression!` lets us avoid the combinatorial explosion of REG/RSV/CapacityReserveMargin. The same technique should be applied to other reserves formulations.

**Review by lbonaldo — COMMENTED** [review]
Thanks @cfe316. This PR is an excellent refactoring. I checked the logic, and it looks correct to me. I added a few comments to the code that are mostly tiny fixes.

File: src/model/expression_manipulation.jl [hunk 1]
```diff
@@ -141,4 +167,4 @@ end
 
 function sum_expression(expr::AbstractArray{AbstractJuMPScalar, dims}) where {dims}
     return _sum_expression(expr)
-end
\ No newline at end of file
+end
```
[the remark this answers is not in the document]
Comment (cfe316) [reply src/model/expression_manipulation.jl:170]: That's weird, I don't know.

File: src/model/expression_manipulation.jl [hunk 2]
```diff
@@ -47,6 +47,32 @@ function fill_with_const!(arr::Array{GenericAffExpr{C,T}, dims}, con::Real) wher
...
+function extract_time_series_to_expression(var::JuMP.Containers.DenseAxisArray{VariableRef, 2, Tuple{X, Base.OneTo{Int64}}, Y},
+                                           set::AbstractVector{Int}) where {X, Y}
+    TIME_DIM = 2
+    time_range = var.axes[TIME_DIM]
+
+    aff_exprs = AffExpr.(0, var[set, :] .=> 1)
+    new_axes = (set, time_range)
+    expr = JuMP.Containers.DenseAxisArray(aff_exprs.data, new_axes...)
+    return expr
+end
```
[the remark this answers is not in the document]
Comment (cfe316) [reply src/model/expression_manipulation.jl:75]: I've added tests now.

**Comment by cfe316** [comment]
Here are the results of the refactor:

Costs / 1e9 for the RealSystemExample/ISONE_Trizone case.

```
5.865087 no reserves

5.883106 original
5.883112 refactored
```
I then increased `Reg_Req_Percent_Demand` and `Rsv_Req_Percent_Demand` to 0.1 (10x and 3x increases).
```
6.281472 original
6.281460 refactored
```
These numbers are very similar; I'd say this was a success.

**Comment by lbonaldo** [comment]
> Here are the results of the refactor:
> 
> Costs / 1e9 for the RealSystemExample/ISONE_Trizone case, which includes THERM, HYDRO, VRE, MUST_RUN, and STOR generators.
> 
> ```
> 5.865087 Reserves=0
> 
> Reserves=1
> 5.883106 original
> 5.883112 refactored
> ```
> 
> I then increased `Reg_Req_Percent_Demand` and `Rsv_Req_Percent_Demand` to 0.1 (10x and 3x increases).
> 
> ```
> 6.281472 original
> 6.281460 refactored
> ```
> 
> These numbers are very similar; I'd say this was a success.

@cfe316 Thank you for the new commits! Could you please also share the settings of the solver you used to run the example?

**Review by lbonaldo — APPROVED** [review]

**Pull request merged**